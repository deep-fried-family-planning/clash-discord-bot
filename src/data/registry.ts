import {ClanVerification, PlayerVerification} from '#src/data/util/index.ts';
import * as GSI2 from '#src/data/gsi/gsi2.ts';
import * as Clan from '#src/data/partition-server/clan.ts';
import * as Server from '#src/data/partition-server/server.ts';
import * as Player from '#src/data/partition-user/player.ts';
import * as UserPartition from '#src/data/partition-user/user.partition.ts';
import * as User from '#src/data/partition-user/user.ts';
import {ClashOfClans} from '#src/service/ClashOfClans.ts';
import * as Array from 'effect/Array';
import type * as Cause from 'effect/Cause';
import * as Data from 'effect/Data';
import type * as DateTime from 'effect/DateTime';
import * as E from 'effect/Effect';
import {pipe} from 'effect/Function';
import * as Order from 'effect/Order';
import * as Record from 'effect/Record';

export class RegistryDefect extends Data.TaggedError('RegistryDefect')<{
  cause: string;
}> {}

export class RegistryFailure extends Data.TaggedError('RegistryFailure')<{
  cause: Cause.Cause<Error>;
}> {}

export class RegistryUserError extends Data.TaggedError('RegistryUserError')<{
  message: string;
}> {}

export class RegistryAdminError extends Data.TaggedError('RegistryAdminError')<{
  message: string;
}> {}

export class RegistryAssertError extends Data.TaggedError('RegistryAssertError')<{
  message: string;
}> {}

const saveUser = (user: Parameters<typeof User.make>[0]) =>
  pipe(
    User.create(User.make(user)),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
  );

const getUser = (user_id: string) =>
  pipe(
    User.read({
      Key           : {pk: user_id, sk: '@'},
      ConsistentRead: true,
    }),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
    E.map((res) => res.Item),
  );

export const assertUser = (user_id: string) =>
  pipe(
    getUser(user_id),
    E.flatMap(E.fromNullable),
    E.catchTag('NoSuchElementException', () => new RegistryAssertError({message: 'User is not registered.'})),
  );

const saveServer = (server: Parameters<typeof Server.make>[0]) =>
  pipe(
    Server.create(Server.make(server)),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
  );

const getServer = (server_id: string) =>
  pipe(
    Server.read({
      Key           : {pk: server_id, sk: '@'},
      ConsistentRead: true,
    }),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
    E.map((res) => res.Item),
  );

export const assertServer = (server_id: string) =>
  pipe(
    getServer(server_id),
    E.flatMap(E.fromNullable),
    E.catchTag('NoSuchElementException', () => new RegistryAssertError({message: 'User is not registered.'})),
  );

const savePlayer = (player: Parameters<typeof Player.make>[0]) =>
  pipe(
    Player.create(Player.make(player)),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
  );

const getPlayer = (user_id: string, player_tag: string) =>
  pipe(
    Player.read({
      Key           : {pk: user_id, sk: player_tag},
      ConsistentRead: true,
    }),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
    E.map((res) => res.Item),
  );

export const assertPlayer = (user_id: string, player_tag: string) =>
  pipe(
    getPlayer(user_id, player_tag),
    E.flatMap(E.fromNullable),
    E.catchTag('NoSuchElementException', () => new RegistryAssertError({message: 'Player is not registered.'})),
  );

const saveClan = (clan: Parameters<typeof Clan.make>[0]) =>
  pipe(
    Clan.create(Clan.make(clan)),
    E.catchAllCause((cause) => new RegistryFailure({cause})),
  );

type Caller = {
  id    : string;
  roles?: string[];
};

type CallerContext = {
  id: string;
};

type UserParams = {
  caller_id : string;
  target_id?: string | undefined;
  payload: {
    timezone: DateTime.TimeZone;
  };
};

export const registerUser = E.fn('registerUser')(function* (params: UserParams) {
  if (params.target_id === params.caller_id) {
    return yield* new RegistryAdminError({
      message: 'You cannot admin re-register yourself.',
    });
  }

  const caller = yield* getUser(params.caller_id);

  if (params.target_id) {
    if (!caller) {
      return yield* new RegistryAdminError({
        message: 'Caller is not registered.',
      });
    }

    const target = yield* getUser(params.target_id);

    if (target) {
      return yield* new RegistryAdminError({
        message: 'User already registered.',
      });
    }

    yield* saveUser({
      pk     : params.target_id,
      sk     : '@',
      pk1    : params.target_id,
      servers: new Set([]),
      ...params.payload,
    });

    return {
      description: 'Success',
    };
  }

  if (!caller) {
    yield* User.create(
      User.make({
        pk     : params.caller_id,
        sk     : '@',
        pk1    : params.caller_id,
        servers: new Set([]),
        ...params.payload,
      }),
    );

    return {
      description: 'Success',
    };
  }

  const updated = User.make({
    ...caller,
    ...params.payload,
  });

  if (!User.equal(updated, caller)) {
    yield* User.create(updated);
  }

  return {
    description: 'Success',
  };
});

type ServerParams = {
  caller_id   : string;
  caller_roles: string[];
  guild_id    : string;
  payload: {
    admin: string;
  };
};

export const registerServer = E.fn('registerServer')(function* (params: ServerParams) {
  if (!params.caller_roles.includes(params.payload.admin)) {
    return yield* new RegistryUserError({
      message: 'You do not have the given admin role.',
    });
  }

  const server = yield* getServer(params.guild_id);

  if (!server) {
    yield* saveServer({
      pk : params.guild_id,
      sk : '@',
      pk1: params.guild_id,
      ...params.payload,
    });

    return {
      description: 'Success',
    };
  }

  if (!params.caller_roles.includes(server.admin)) {
    return yield* new RegistryAdminError({
      message: 'You do not have the current admin role to update this server registration.',
    });
  }

  yield* saveServer({
    ...server,
    pk: params.guild_id,
    sk: '@',
    ...params.payload,
  });

  return {
    description: 'Success',
  };
});

type ClanParams = {
  caller_id   : string;
  caller_roles: string[];
  guild_id    : string;
  clan_tag    : string;
  payload: {
    countdown: string;
  };
};

export const registerClan = E.fn('registerClan')(function* (p: ClanParams) {
  const server = yield* assertServer(p.guild_id);

  if (!p.caller_roles.includes(server.admin)) {
    return yield* new RegistryAdminError({
      message: 'You are not authorized to register a clan for this server.',
    });
  }

  const userPartition = yield* UserPartition.scanUp(p.caller_id);

  const user = userPartition.Items.find((u) => User.is(u));

  if (!user) {
    return yield* new RegistryAdminError({
      message: 'Your Discord user account is not registered.',
    });
  }

  const userPlayers = pipe(
    userPartition.Items,
    Array.filter((u) => Player.is(u)),
    Record.fromIterableWith((up) => [up.sk, up]),
  );

  if (!Record.size(userPlayers)) {
    return yield* new RegistryAdminError({
      message: 'You have no player accounts registered to your Discord account.',
    });
  }

  const gsi = yield* GSI2.queryClan(p.clan_tag);

  if (gsi.Items.length > 1) {
    return yield* new RegistryDefect({
      cause: 'Multiple clan records found.',
    });
  }

  const clan = yield* ClashOfClans.getClan(p.clan_tag);

  const verifications = pipe(
    clan.members,
    Array.filter((m) => m.role !== 'member'),
    Array.filter((m) => m.tag in userPlayers),
    Array.filter((m) => userPlayers[m.tag].verification >= PlayerVerification.token),
    Array.map((m) =>
      m.role === 'leader' ? ClanVerification.leader :
        m.role === 'coLeader' ? ClanVerification.coleader :
          ClanVerification.elder,
    ),
    Array.sort(Order.number),
  );

  if (!verifications.length) {
    return yield* new RegistryAdminError({
      message: 'Your Discord account has no verified player links (elder or above) in this clan.',
    });
  }

  const verification = verifications[0];

  if (gsi.Items.length === 1) {
    const current = gsi.Items[0];

    if (current.verification > verification) {
      return yield* new RegistryAdminError({
        message: 'Your member player links cannot override the current verification level of this clan.',
      });
    }

    if (current.pk !== p.guild_id) {
      yield* Clan.remove({
        Key: {pk: current.pk, sk: current.sk},
      });

      yield* saveClan({
        pk         : p.guild_id,
        sk         : p.clan_tag,
        pk2        : p.clan_tag,
        sk2        : p.guild_id,
        name       : clan.name,
        description: clan.description,
        select     : {
          value: p.clan_tag,
          label: clan.name,
        },
        verification,
        ...p.payload,
      });
    }
    else {
      yield* saveClan({
        ...current,
        name       : clan.name,
        description: clan.description,
        verification,
        ...p.payload,
      });
    }

    return {
      description: 'Success',
    };
  }

  yield* saveClan({
    pk         : p.guild_id,
    sk         : p.clan_tag,
    pk2        : p.clan_tag,
    sk2        : p.guild_id,
    name       : clan.name,
    description: clan.description,
    select     : {
      value: p.clan_tag,
      label: clan.name,
    },
    verification,
    ...p.payload,
  });

  return {
    description: 'Success',
  };
});

type PlayerParams = {
  guild_id?    : string;
  caller_id    : string;
  caller_roles?: string[];
  player_tag   : string;
  api_token?   : string | undefined;
  target_id?   : string | undefined;
  payload: {
    account_type: string;
  };
};

export const registerPlayer = E.fn('registerPlayer')(function* (p: PlayerParams) {
  yield* assertUser(p.caller_id);

  if (!p.api_token) {
    if (!p.target_id) {
      return yield* new RegistryAdminError({
        message: 'You must provide a target user id to admin register a player account.',
      });
    }
    if (!p.guild_id || !p.caller_roles) {
      return yield* new RegistryAdminError({
        message: 'Admin registration must be within context of a server.',
      });
    }

    const server = yield* assertServer(p.guild_id);

    if (!p.caller_roles.includes(server.admin)) {
      return yield* new RegistryAdminError({
        message: 'You are not authorized to register a player account for this server.',
      });
    }

    const gsi = yield* GSI2.queryPlayer(p.player_tag);

    if (gsi.Items.length > 1) {
      return yield* new RegistryDefect({cause: 'Multiple player records found.'});
    }

    const player = yield* ClashOfClans.getPlayer(p.player_tag);

    if (gsi.Items.length === 1) {
      const current = gsi.Items[0];

      if (current.verification >= PlayerVerification.token) {
        return yield* new RegistryAdminError({
          message: 'You are not authorized to update the registration of this player account.',
        });
      }

      yield* Player.remove({
        Key: {
          pk: current.pk,
          sk: current.sk,
        },
      });

      yield* savePlayer({
        pk          : p.target_id,
        sk          : p.player_tag,
        pk2         : p.player_tag,
        sk2         : p.target_id,
        name        : player.name,
        ...p.payload,
        verification: PlayerVerification.admin,
      });

      return {
        description: 'Success',
      };
    }

    yield* Player.create(
      Player.make({
        pk          : p.target_id,
        sk          : p.player_tag,
        pk2         : p.player_tag,
        sk2         : p.target_id,
        name        : player.name,
        ...p.payload,
        verification: PlayerVerification.admin,
      }),
    );

    return {
      description: 'Success',
    };
  }

  const isValid = yield* ClashOfClans.verifyPlayerToken(p.player_tag, p.api_token);

  if (!isValid) {
    return yield* new RegistryUserError({
      message: 'Invalid player token.',
    });
  }

  const gsi = yield* GSI2.queryPlayer(p.player_tag);

  if (gsi.Items.length > 1) {
    return yield* new RegistryDefect({
      cause: 'Multiple player records found.',
    });
  }

  const player = yield* ClashOfClans.getPlayer(p.player_tag);

  if (gsi.Items.length === 1) {
    const current = gsi.Items[0];

    if (current.pk !== p.caller_id) {
      yield* Player.remove({
        Key: {
          pk: current.pk,
          sk: current.sk,
        },
      });

      yield* Player.create(
        Player.make({
          ...current,
          pk          : p.caller_id,
          sk          : p.player_tag,
          pk2         : p.player_tag,
          sk2         : p.caller_id,
          name        : player.name,
          ...p.payload,
          verification: PlayerVerification.token,
        }),
      );

      return {
        description: 'Success',
      };
    }

    yield* Player.create(
      Player.make({
        ...current,
        name        : player.name,
        ...p.payload,
        verification: PlayerVerification.token,
      }),
    );

    return {
      description: 'Success',
    };
  }

  yield* Player.create(
    Player.make({
      pk          : p.caller_id,
      sk          : p.player_tag,
      pk2         : p.player_tag,
      sk2         : p.caller_id,
      name        : player.name,
      ...p.payload,
      verification: PlayerVerification.token,
    }),
  );

  return {
    description: 'Success',
  };
});
