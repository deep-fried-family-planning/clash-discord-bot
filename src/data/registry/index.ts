export * as UserRegistry from './user-registry.ts';
export type UserRegistry = typeof UserRegistry;
