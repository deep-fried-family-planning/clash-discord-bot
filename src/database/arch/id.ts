import {S} from '#src/internal/pure/effect.ts';
import {pipe} from 'effect';

const asPrependedId = (start: string) =>
  pipe(
    S.String,
    S.startsWith(start),
    S.transform(
      S.String,
      {
        decode: (s) => s.slice(start.length),
        encode: (s) => start.concat(s),
      },
    ),
    S.transform(
      S.String,
      {
        decode: (s) => {
          if (s.startsWith(start)) {
            throw new Error();
          }
          return s;
        },
        encode: (s) => {
          if (s.startsWith(start)) {
            throw new Error();
          }
          return s;
        },
      },
    ),
  );

export const ClanTag = asPrependedId('c-');
export const PlayerTag = asPrependedId('p-');
export const ServerId = asPrependedId('s-');
export const UserId = asPrependedId('u-');
export const ThreadId = S.String;
export const RoleId = S.String;
export const ChannelId = S.String;
export const AllianceId = asPrependedId('a-');
export const InfoId = asPrependedId('i-');
export const EmbedId = asPrependedId('e-');
export const NowSk = S.Literal('now');
export const TransientUrl = S.String.pipe(S.length({min: 1, max: 2000}));
export const ComputeName = S.String;
export const DateTimeSk = S.DateTimeUtc;

export * as Id from '#src/database/arch/id.ts';
export type Id = | typeof ClanTag
                 | typeof PlayerTag
                 | typeof ServerId
                 | typeof UserId
                 | typeof ThreadId
                 | typeof RoleId
                 | typeof ChannelId
                 | typeof AllianceId
                 | typeof InfoId
                 | typeof EmbedId
                 | typeof NowSk
                 | typeof TransientUrl
                 | typeof ComputeName
                 | typeof DateTimeSk;
