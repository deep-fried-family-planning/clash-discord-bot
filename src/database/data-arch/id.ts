import {asPrependedId} from '#src/database/data-arch/codec-standard.ts';
import {S} from '#src/internal/pure/effect.ts';

export const ClanTag = asPrependedId('c-');
export const PlayerTag = asPrependedId('p-');

export const ServerId = asPrependedId('s-');
export const UserId = asPrependedId('u-');
export const ThreadId = S.String;
export const RoleId = S.String;
export const ChannelId = S.String;

export const AllianceId = asPrependedId('a-');
export const InfoId = asPrependedId('i-');
export const EmbedId = asPrependedId('e-');
export const NowSk = S.Literal('now');

export const TransientUrl = S.String.pipe(S.length({min: 1, max: 2000}));
export const ComputeName = S.String;
export const DateTimeSk = S.DateTimeUtc;

export * as Id from '#src/database/data-arch/id.ts';
export type Id = | typeof ClanTag
                 | typeof PlayerTag
                 | typeof ServerId
                 | typeof UserId
                 | typeof ThreadId
                 | typeof RoleId
                 | typeof ChannelId
                 | typeof AllianceId
                 | typeof InfoId
                 | typeof EmbedId
                 | typeof NowSk
                 | typeof TransientUrl
                 | typeof ComputeName
                 | typeof DateTimeSk;
