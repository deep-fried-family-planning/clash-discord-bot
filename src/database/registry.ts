import {ClashOfClans} from '#src/clash/clashofclans.ts';
import {UserPlayer} from '#src/database/data/codec.ts';
import {deleteItem, queryUserPlayers, saveItem} from '#src/database/db.ts';
import {COLOR, nColor} from '#src/internal/discord-old/constants/colors.ts';
import {dLinesS} from '#src/internal/discord-old/util/markdown.ts';
import {E} from '#src/internal/pure/effect.ts';
import type {Embed} from 'dfx/types';
import {Data} from 'effect';

export class RegistrationError extends Data.TaggedError('deepfryer/RegistrationError')<{
  issue: string;
}> {}

export class RegistrationDefect extends Data.TaggedError('deepfryer/RegistrationDefect')<{
  issue: string;
}> {}

export const registerServer = () => E.gen(function* () {

});

export const updateRegisteredServer = () => E.gen(function* () {

});

export const registerServerClan = () => E.gen(function* () {

});

export const updateRegisteredServerClan = () => E.gen(function* () {

});

export const registerUser = (ops: {}) => E.gen(function* () {

});

export const updateRegisteredUser = (ops: {}) => E.gen(function* () {

});

export const registerUserPlayers = (ops: {}) => E.gen(function* () {

});

type UserPlayerRegistrationOps = {
  user    : string;
  roles   : string[];
  tag     : string;
  apiToken: string;
  kind    : string;
  admin   : string;
  target? : string;
};

export const registerUserPlayer = (ops: UserPlayerRegistrationOps) => E.gen(function* () {
  const tag = yield* ClashOfClans.validateTag(ops.tag);
  const isAdmin = ops.roles.includes(ops.admin);

  if (ops.apiToken === 'admin') {
    if (!isAdmin) {
      return yield* new RegistrationError({issue: 'admin role required'});
    }
    if (!ops.target) {
      return yield* new RegistrationError({issue: 'admin links must have discord_user'});
    }

    const player = yield* ClashOfClans.getPlayer(tag);
    const [dbPlayer, ...rest] = yield* queryUserPlayers(ops.tag);

    if (rest.length) {
      return yield* new RegistrationDefect({issue: 'Index Defect'});
    }

    if (!dbPlayer) {
      const item = UserPlayer.make({
        pk            : ops.target,
        sk            : ops.tag,
        gsi_user_id   : ops.target,
        gsi_player_tag: ops.tag,
        name          : player.name,
        verification  : 1,
        account_type  : ops.kind,
      });

      yield* saveItem(UserPlayer, item);

      return {
        color      : nColor(COLOR.SUCCESS),
        description: dLinesS(
          'New admin link',
          `<@${item.pk}> linked to ${item.name} (${item.sk}) (${item.account_type})`,
        ),
      } as Embed;
    }

    yield* deleteItem(
      UserPlayer,
      dbPlayer.pk,
      dbPlayer.sk,
    );

    const item = UserPlayer.make({
      pk            : ops.target,
      sk            : ops.tag,
      gsi_user_id   : ops.target,
      gsi_player_tag: ops.tag,
      name          : player.name,
      verification  : 1,
      account_type  : ops.kind,
    });

    yield* saveItem(UserPlayer, item);

    return {
      color      : nColor(COLOR.SUCCESS),
      description: dLinesS(
        'Admin override link',
        `<@${item.pk}> linked to ${item.name} (${item.sk}) (${item.account_type})`,
      ),
    } as Embed;
  }

  const isValid = yield* ClashOfClans.verifyPlayerToken(ops.tag, ops.apiToken);

  if (!isValid) {
    return yield* new RegistrationError({
      issue: 'Invalid api_token.',
    });
  }

  const player = yield* ClashOfClans.getPlayer(tag);
  const [dbPlayer, ...rest] = yield* queryUserPlayers(ops.tag);

  if (!dbPlayer) {
    const item = UserPlayer.make({
      pk            : ops.user,
      sk            : ops.tag,
      gsi_user_id   : ops.user,
      gsi_player_tag: ops.tag,
      name          : player.name,
      verification  : 2,
      account_type  : ops.kind,
    });

    yield* saveItem(UserPlayer, item);

    return {
      color      : nColor(COLOR.SUCCESS),
      description: dLinesS(
        'New player link verified',
        `<@${item.pk}> linked to ${item.name} (${item.sk}) (${item.account_type})`,
      ),
    } as Embed;
  }

  if (rest.length) {
    return yield* new RegistrationDefect({issue: 'Index Defect'});
  }
  if (dbPlayer.sk === player.tag) {
    return yield* new RegistrationError({
      issue: 'Account is already linked',
    });
  }
  if (dbPlayer.verification > 2) {
    return yield* new RegistrationError({
      issue: 'You cannot override this account link',
    });
  }

  yield* deleteItem(
    UserPlayer,
    dbPlayer.pk,
    dbPlayer.sk,
  );

  const item = UserPlayer.make({
    pk            : ops.user,
    sk            : ops.tag,
    gsi_user_id   : ops.user,
    gsi_player_tag: ops.tag,
    name          : player.name,
    verification  : 2,
    account_type  : ops.kind,
  });

  yield* saveItem(UserPlayer, item);

  return {
    color      : nColor(COLOR.SUCCESS),
    description: dLinesS(
      'Player link overridden.',
      `<@${item.pk}> linked to ${item.name} (${item.sk}) (${item.account_type})`,
    ),
  } as Embed;
});
