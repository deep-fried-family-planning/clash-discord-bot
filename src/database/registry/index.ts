import * as clanRegistry from './clan-registry.ts';
export type ClanRegistry = never;
export {clanRegistry as ClanRegistry};
