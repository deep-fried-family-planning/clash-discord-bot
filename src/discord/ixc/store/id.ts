import {inject} from 'regexparam';
import type {str} from '#src/internal/pure/types-pure.ts';
import {pipe} from '#src/internal/pure/effect.ts';
import {filterL, reduceL} from '#src/internal/pure/pure-list.ts';
import {emptyKV, keysKv} from '#src/internal/pure/pure-kv.ts';
import type {RDXT} from '#src/discord/ixc/store/types.ts';
import {RDXK} from '#src/discord/ixc/store/types.ts';
import {ID_ROUTES, type Route, type RouteParams, DELIM} from '#src/discord/ixc/store/id-routes.ts';


export const parseCustomId = (custom_id: str) => {
    const route = ID_ROUTES.find((r) => r.pattern.test(custom_id));

    if (!route) {
        return {
            custom_id,
            template: '',
            params  : {
                kind: RDXK.CLOSE,
                data: [],
            },
            predicate    : RDXK.CLOSE,
            nextPredicate: RDXK.CLOSE,
        } as const satisfies Route;
    }

    const [, ...matches] = route.pattern.exec(custom_id)!;

    const params = pipe(
        route.keys,
        reduceL(emptyKV<string, string>(), (ks, k, idx) => {
            ks[k] ??= matches[idx];
            return ks;
        }),
    );

    return {
        custom_id,
        template: route.template,
        params  : {
            kind: params.kind as RDXK,
            type: params.type as RDXT,
            data: 'data' in params
                ? params.data.split(DELIM.DATA).map((d) => d.replaceAll(DELIM.PIPE, DELIM.SLASH))
                : [],
            nextKind: params.nextKind as RDXK,
            nextType: params.nextType as RDXT,
            forward : params.forward,
        },
        predicate    : `${params.kind}/${params.type}`,
        nextPredicate: `${params.nextKind}/${params.nextType}`,
    } as const satisfies Route;
};


export const buildCustomId = (params: RouteParams) => {
    const redefined = {
        ...params,
        data: params.data?.map((d) => d.replaceAll(DELIM.SLASH, DELIM.PIPE)).join(DELIM.DATA),
    } as const;

    const defined = pipe(
        redefined,
        Object.entries,
        filterL(([,v]) => Boolean(v)),
        Object.fromEntries,
    ) as Record<str, str>;

    const route = pipe(
        defined,
        keysKv,
        (ks) => ID_ROUTES.find((r) => r.keys.length === ks.length && r.keys.every((k) => ks.includes(k))),
    );

    return {
        custom_id    : inject(route!.template, defined as Omit<RouteParams, 'data'> & {data: string}),
        template     : route!.template,
        params,
        predicate    : `${params.kind}/${params.type}`,
        nextPredicate: `${params.nextKind}/${params.nextType}`,
    } as const satisfies Route;
};
