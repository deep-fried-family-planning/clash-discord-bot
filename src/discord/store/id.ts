import type {str} from '#src/internal/pure/types-pure.ts';



export const idPredicate = (k?: str, t?: str) => `${k}/${t}`;
