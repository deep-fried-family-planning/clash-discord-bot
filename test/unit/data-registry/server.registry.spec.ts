import * as ServerRegistry from '#src/data-registry/server.registry.ts';
import {TestDataServer} from '#unit/.mock/mock-db.testdata.ts';
import {mockDb, mockDbLayer} from '#unit/.mock/mock-db.ts';
import {it} from '@effect/vitest';
import {DateTimes, GetRandomValues} from '@typed/id';
import * as E from 'effect/Effect';
import {pipe} from 'effect/Function';

it.effect('when registering a new server', E.fn(function* () {
  mockDb.get.mockReturnValueOnce(E.succeed({Item: undefined}));

  const actual = yield* pipe(
    ServerRegistry.register({
      caller_id   : 'user',
      caller_roles: ['admin'],
      guild_id    : 'guild',
      payload     : {
        admin: 'admin',
      },
    }),
    E.provide(mockDbLayer),
    E.provide([GetRandomValues.layer(() => E.succeed(new Uint8Array([]))), DateTimes.Fixed(new Date(0))]),
  );

  expect(actual).toMatchSnapshot();
  expect(mockDb.get.mock.calls[0][0].Key).toMatchSnapshot();
  expect(mockDb.put.mock.calls[0][0].Item).toMatchSnapshot();
}));

it.effect('when re-registering a server', E.fn(function* () {
  mockDb.get.mockReturnValueOnce(E.succeed({Item: TestDataServer}));

  const actual = yield* pipe(
    ServerRegistry.register({
      caller_id   : 'user',
      caller_roles: ['admin', 'admin2'],
      guild_id    : 'guild',
      payload     : {
        admin: 'admin2',
      },
    }),
    E.provide(mockDbLayer),
    E.provide([GetRandomValues.layer(() => E.succeed(new Uint8Array([]))), DateTimes.Fixed(new Date(0))]),
  );

  expect(actual).toMatchSnapshot();
  expect(mockDb.get.mock.calls[0][0].Key).toMatchSnapshot();
  expect(mockDb.put.mock.calls[0][0].Item).toMatchSnapshot();
}));
