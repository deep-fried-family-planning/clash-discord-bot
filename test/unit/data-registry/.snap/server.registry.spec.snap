// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when re-registering a server 1`] = `
{
  "description": "Success",
}
`;

exports[`when re-registering a server 2`] = `
{
  "pk": "guild",
  "sk": "@",
}
`;

exports[`when re-registering a server 3`] = `
{
  "_tag": "Server",
  "_v": 1,
  "_v7": "00000000-0000-7000-8000-000000000000",
  "admin": "admin2",
  "created": "1970-01-01T00:00:00.000Z",
  "pk": "guild",
  "pk1": "guild",
  "sk": "@",
  "sk1": "@",
  "updated": "1970-01-01T00:00:00.000Z",
}
`;

exports[`when registering a new server 1`] = `
{
  "description": "Success",
}
`;

exports[`when registering a new server 2`] = `
{
  "pk": "guild",
  "sk": "@",
}
`;

exports[`when registering a new server 3`] = `
{
  "_tag": "Server",
  "_v": 1,
  "_v7": "00000000-0000-7000-8000-000000000000",
  "admin": "admin",
  "created": "1970-01-01T00:00:00.000Z",
  "pk": "guild",
  "pk1": "guild",
  "sk": "@",
  "sk1": "@",
  "updated": "1970-01-01T00:00:00.000Z",
}
`;
